### PR TITLE
[MERGE] hr_timesheet: Improve Timer, Access Rights and Validation

### PR DESCRIPTION
**Pull request by tivisse**
Access rights
=============

Purpose
-------

Currently, the lowest access right level for Timesheets allows a user
to see the timesheets from all employees. Some companies do not want
to have this information shared.

As we already have three levels of access rights, creating a new one
would be too much. Instead, we are using this opportunity to
clean/simplify the current access right levels.

Specification
-------------

Remove the current access rights and replace them with the following:
- 'See own timesheets' the user can:
    - read/write/create/delete his own timesheets only
    - cannot timesheet on projects whose Visibility is set on
      'Invited employees' if the user is not a follower
    - see the 'My Timesheets' and the 'Timesheet App' menus
- 'See all timesheets' the user can:
    - read/write/create/delete all timesheets
    - cannot see the Timesheets of Projects whose Visibility is set
      on 'Invited employees' if the user is not a follower
    - see the 'My Timesheets' + 'All Timesheets' + 'Timesheet App' +
      'To Validate' menus
    - validate the timesheets of whom he is the 'Timesheet
      Responsible' or the 'Manager'
- 'Administrator' the user can:
    - read/write/create/delete all timesheets
    - see all menus
    - validate the Timesheets of all employees
- Write tests

New Validation Process
======================

Purpose
-------

Currently, timesheets are validated by employees as a whole. However,
as a manager, it is difficult to evaluate the timesheets of an
employee for projects you know nothing about. Therefore, it would make
more sense for a manager to validate the timesheets of his projects
and to leave the rest to the other managers.

In addition, having a date on the employees leaves room for error and
performance issues, so we are introducing a new checkbox at the
timesheet level.

Specification
-------------

Remove the 'Timesheets Validation Date' field on the employee
- replace it with a 'Timesheet Validated' checkbox (hidden field) on
  the timesheet entry
    - By default, this field is set to false
    - Once the manager validates the timesheets, switch the field to
      true
- Filtering on 'Validated' timesheets should display timesheets for
  which 'Timesheet Validated' is true
- Filtering on 'Non-Validated' timesheets should display timesheets
  for which 'Timesheet Validated' is false

Rename the title of the validation menus into:
- To Validate > Last Weeks -> Timesheets from Last Week to Validate
- To Validate > Last Month -> Timesheets from Last Month to Validate

Grid view: display projects on which the user timesheeted in the
past 30 days (group_expand), or the employees who timesheeted in the
past 30 days for the 'To Validate' menus)

Open the 'To Validate' menus on the oldest non-validated timesheet
you can find.

Clicking on 'Validate' should validate all the timesheets displayed
on the screen. List all the elements displayed in the view based on
the group by applied in the 'Validate the timesheets' modal.

When validating timesheets, bypass timesheets that were already
validated in the past in order to avoid performance issues

Timer list
==========

Purpose
-------

Have an overview of the timesheets you need to do during the day.
Easily track time spent on tasks using the timer instead of having to
write down somewhere at what time you started an activity. Allow more
flexibility if an activity if interrupted by another one (so you
don't need to do mental arithmetic to determine how much time the
first activity took substracted by the time spent on the second
activity).

Specification
-------------

- Move the time from fsm to hr_timesheet.
- Add fa-play-circle and fa-stop-circle icons in the 'My Timesheets' 
  menu:
    - on the left of the 'Duration' field in the list view
    - on the top right corner of the kanban card in the kanban view
    - on the right of the first column in the grid view
- Display these icons
    - for non-validated timesheets only
    - if the Encoding Unit is Hours
- Clicking on the play button should start and display the timer and
   switch to the stop icon
- Clicking on the stop button should halt and stop displaying the
  timer and switch to the play icon
- While the timer is running, the 'Duration' field should not
  be editable
- Only one timer can run at the same time. If the user clicks on the
  start button of timesheet B while the timer is running for
  timesheet A, timesheet B should start and timesheet A should stop.

Time selection when invoicing
=============================

Purpose
-------

Allow selecting a timeframe to invoice when the user is invoicing on
a periodical basis. For instance, the user invoices each time at the
end of the month. For some reason (he got sick, there was a bank
holiday...), he only invoices of the 4th of the following month.
Now the timesheets that are about to be invoiced include some that
were recorded in the meantime.

Specification
-------------

- Add an 'Invoice timesheets up to mm/dd/yyyy' field on the Invoice
  Creation modal
    - display this field only when creating an Invoice for a Service
      whose Invoicing Policy is on 'Timesheets on tasks'
    - not required
    - Set today's date by default
    - if this field is set, the quantity to invoice should only
      include the not yet invoiced (and validated) timesheets up to
      this date (included)
    - if this field is not set, include all not yet invoiced (and
      validated) timesheets without distinction.
    - Add the following tooltip: Only timesheets not yet invoiced
      (and validated, if applicable) up to this date included will be
      invoiced. If no date is indicated, all timesheets not yet
      invoiced (and validated, if applicable) will be invoiced
      without distinction.

Miscellaneous
=============

Purpose
-------

Small usability improvements.

- Move 'Create SO from Task' to sale_timesheet -> Means Enterprise to
  Community).
- Add the pivot view to the 'All Timesheets' menu
- Add the 'Today' button on the grid view of all menus (see the
  Reporting > Timesheets by Project menu for reference)

Co-authored-by: Lucas Lefèvre <lul@odoo.com>
Co-authored-by: Yannick Tivisse <yti@odoo.com>

TaskID: 2088586

### CODE REVIEW

**Comment by mreficent** [comment]
@alexey-pelykh @astirpe may you be interested in this PR.

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 1]
```diff
@@ -84,6 +99,12 @@ def create(self, values):
         return result
 
     def write(self, values):
+        # If it's a basic user then check if the timesheet is his own.
```
Comment (alexey-pelykh) [inline addons/hr_timesheet/models/hr_timesheet.py]: Ideally, this should be separate PR?

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 2]
```diff
@@ -69,6 +78,12 @@ def _compute_department_id(self):
 
     @api.model
     def create(self, values):
+        # If it's a basic user then check if the timesheet is his own.
```
Comment (alexey-pelykh) [inline addons/hr_timesheet/models/hr_timesheet.py]: Ideally, this should be separate PR?

File: addons/hr_timesheet/models/hr_timesheet.py (hunk 1 above)
Comment (alexey-pelykh) [reply addons/hr_timesheet/models/hr_timesheet.py]: + It would be great to make separate permissions to alter records of another user. Also, such implementation will limit extensibility for other modules?

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 3]
```diff
@@ -180,3 +201,72 @@ def _timesheet_postprocess_values(self, values):
...
+            unit_amount based on the timer and the previous value of
+            unit_amount for the current timesheet.
+        """
+        if self.timer_start and self.display_timer:
+            precision = self.env['decimal.precision'].precision_get('Product Unit of Measure')
+            minutes_spent = self._compute_minutes_spent()
+
+            if float_is_zero(self.unit_amount, precision_digits=precision) and float_is_zero(minutes_spent, precision_digits=precision):
+                # Check if unit_amount and minutes_spent is not 0,
+                # if yes, we set minutes_spent to 1.
```
Comment (LucasLefevre) [inline addons/hr_timesheet/models/hr_timesheet.py]: Why?

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 4]
```diff
@@ -180,3 +201,72 @@ def _timesheet_postprocess_values(self, values):
...
+            })
+
+    def _search_timer_start(self):
+        """ Search if a timesheet has a timer activated and stop the timer.
+
+            Check if a timer is activated for another timesheet
+            if yes, then update unit_amount field and stop timer,
+            otherwise, do nothing.
+        """
+        record = self.env['account.analytic.line'].search([
```
Comment (LucasLefevre) [inline addons/hr_timesheet/models/hr_timesheet.py]: `record = self.search(...)`
self is already an 'account.analytic.line'

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 5]
```diff
@@ -69,6 +69,11 @@ def _compute_department_id(self):
 
     @api.model
     def create(self, values):
+        # If it's a basic user then check if the timesheet is his own.
+        if not self.user_has_groups('hr_timesheet.group_hr_timesheet_approver'):
+            if self.env.user.id != values['user_id']:
```
Comment (tivisse) [inline addons/hr_timesheet/models/hr_timesheet.py]: user_id isn't required, even if it has a default value. So ``values['user_id']`` could crash.

You can do something like
``` python
if 'user_id' in values and self.env.user.id != values['user_id'] and not self.user_has_groups('hr_timesheet.group_hr_timesheet_approver'):
    raise AccessError(_("You aren't allowed to access to the timesheet that it's not yours"))
```

On the other hand, the message doesn't say obviously what's wrong. You can raise something like: "Sorry, you are only allowed to create yourself timesheets."

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 6]
```diff
@@ -84,6 +89,11 @@ def create(self, values):
         return result
 
     def write(self, values):
+        # If it's a basic user then check if the timesheet is his own.
+        if not self.user_has_groups('hr_timesheet.group_hr_timesheet_approver'):
+            if self.env.user.id != self.user_id.id:
+                raise AccessError(_("You aren't allowed to access to the timesheet that it's not yours"))
```
Comment (tivisse) [inline addons/hr_timesheet/models/hr_timesheet.py]: Same here, for both comments.

File: addons/hr_timesheet/security/ir.model.access.csv [hunk 7]
```diff
@@ -1,5 +1,7 @@
 id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
-access_account_analytic_line_user,analytic.account.analytic.line.timesheet.user,analytic.model_account_analytic_line,hr_timesheet.group_hr_timesheet_user,1,1,1,1
+access_account_analytic_line_manager,analytic.account.analytic.line.timesheet.manager,analytic.model_account_analytic_line,hr_timesheet.group_timesheet_manager,1,1,1,1
+access_account_analytic_line_approver,analytic.account.analytic.line.timesheet.approver,analytic.model_account_analytic_line,hr_timesheet.group_hr_timesheet_approver,1,1,1,1
```
[the remark this answers is not in the document]
Comment (tivisse) [reply addons/hr_timesheet/security/ir.model.access.csv]: Yep, more generally, your should only define the access rules for the lowest user's group to own the privilege.

Example: Grant CRUD access rights to users only, and not to managers and officers and user.

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 8]
```diff
@@ -180,3 +201,72 @@ def _timesheet_postprocess_values(self, values):
...
+            If yes, then stop the timer before launch this timer.
+        """
+        self.ensure_one()
+        if not self.timer_start and self.display_timer:
+            self._search_timer_start()
+            self.write({
+                'timer_start': fields.Datetime.now(),
+            })
+
+    def _search_timer_start(self):
```
[the remark this answers is not in the document]
Comment (tivisse) [reply addons/hr_timesheet/models/hr_timesheet.py]: :+1:

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 9]
```diff
@@ -27,8 +27,11 @@ def default_get(self, field_list):
     department_id = fields.Many2one('hr.department', "Department", compute='_compute_department_id', store=True, compute_sudo=True)
     encoding_uom_id = fields.Many2one('uom.uom', compute='_compute_encoding_uom_id')
 
+    # Timer Fields
     timesheet_timer_start = fields.Datetime("Timesheet Timer Start", default=None)
 
+    timer = fields.Boolean(string="", compute="_compute_timer", store=False, help="This field is used for timer_toggle_button widget")
```
Comment (tivisse) [inline addons/hr_timesheet/models/hr_timesheet.py]: The help message is not very understandable for a common human. Could you adapt ?

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 10]
```diff
@@ -198,7 +201,19 @@ def _timesheet_postprocess_values(self, values):
     # ----------------------------------------------------
     # Timer Methods
     # ----------------------------------------------------
+    @api.depends('timesheet_timer_start')
+    def _compute_timer(self):
+        for record in self:
+            if record.timesheet_timer_start:
+                record.timer = True
+            else:
+                record.timer = False
```
Comment (tivisse) [inline addons/hr_timesheet/models/hr_timesheet.py]: Use a better descriptor than record, for example analytic_line
``` python
    @api.depends('timesheet_timer_start')
    def _compute_timer(self):
        for analytic_line in self:
            analytic_line.timer = True if analytic_line.timesheet_timer_start else False
```

File: addons/hr_timesheet/models/hr_timesheet.py [hunk 11]
```diff
@@ -217,6 +232,11 @@ def _search_timer_start(self):
             record.action_timer_stop()
 
     def action_timer_stop(self):
+        """
+            When the timer must be stopped, we must calculate the new
+            unit_amount based on the timer and the previous value of
+            unit_amount for the current timesheet.
+        """
```
Comment (tivisse) [inline addons/hr_timesheet/models/hr_timesheet.py]: Nice to add docstrings ;)

File: addons/hr_timesheet/views/hr_timesheet_views.xml [hunk 12]
```diff
@@ -21,11 +21,16 @@
...
-                <tree editable="top" string="Timesheet Activities">
+                <tree editable="top" string="Timesheet Activities" js_class="timesheet_tree">
                     <field name="date"/>
                     <field name="name"/>
                     <field name="project_id" required="1" context="{'form_view_ref': 'project.project_project_view_form_simplified',}"/>
                     <field name="task_id" context="{'default_project_id': project_id}" domain="[('project_id', '=', project_id)]"/>
+                    <field
+                        name="timer"
+                        widget="timer_toggle_button"
+                        readonly='True'
```
Comment (tivisse) [inline addons/hr_timesheet/views/hr_timesheet_views.xml]: As truthy as ``readonly='0'``, and yet less readable. You can also inline small attributes like that ;)

As a reminder, we don't apply those pep8 rules (See: https://www.odoo.com/documentation/online/reference/guidelines.html):
E501: line too long
E301: expected 1 blank line, found 0
E302: expected 2 blank lines, found 1
E126: continuation line over-indented for hanging indent
E123: closing bracket does not match indentation of opening bracket's line
E127: continuation line over-indented for visual indent
E128: continuation line under-indented for visual indent
E265: block comment should start with '# '

File: addons/hr_timesheet/views/hr_timesheet_views.xml [hunk 13]
```diff
@@ -161,6 +167,13 @@
                                             <span>
                                                 <t t-esc="record.name.value"/>
                                             </span>
+                                            <div class="btn-action float-right">
+                                                <field
```
Comment (tivisse) [inline addons/hr_timesheet/views/hr_timesheet_views.xml]: Same here, no need for 6 lines here ;)

File: addons/hr_timesheet/views/hr_timesheet_views.xml [hunk 14]
```diff
@@ -31,12 +31,13 @@
                         widget="timer_toggle_button"
                         readonly='True'
                         attrs="{
-                            'invisible': [('encoding_uom_name', '!=', 'Hours')]
+                            'invisible': &quot;['|', ('encoding_uom_name', '!=', 'Hours'), ('user_id', '!=', uid)]&quot;
```
Comment (tivisse) [inline addons/hr_timesheet/views/hr_timesheet_views.xml]: Small question: And if a user can see all the timesheets. Do we cant the same restriction ? Maybe if we want to restrict according to the group, we should introduce a simple computed field (non stored) to display the button or not.

To check with (aju).

File: addons/hr_timesheet/views/hr_timesheet_views.xml [hunk 15]
```diff
@@ -188,6 +210,7 @@
             <field name="view_mode">tree,form</field>
             <field name="domain">[('project_id', '!=', False), ('user_id', '=', uid)]</field>
             <field name="context">{
+                "timer": True,
```
[the remark this answers is not in the document]
Comment (tivisse) [reply addons/hr_timesheet/views/hr_timesheet_views.xml]: 1/ Try to remove the context key
2/ Try to remove it again, if 1/ failed.
3/ Rename it into something more explicit and easy to grep than 'timer'. (Currently 715 matches in master). For example, 'hr_timesheet_display_timer' (if this is what it does).

File: addons/hr_timesheet/models/timer_mixin.py [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+
+    def action_timer_start(self) -> None:
+        """ Action start the timer.
+
+            Start timer and search if another timer hasn't been launched.
+            If yes, then stop the timer before launch this timer.
+        """
+        self.ensure_one()
+        if not self.timer_start:
+            self.write({
```
Comment (tivisse) [inline addons/hr_timesheet/models/timer_mixin.py]: You can inline small chunks of code like that.

File: addons/hr_timesheet/models/timer_mixin.py [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+
+    def action_timer_stop(self):
+        """ Action stop the timer.
+
+            Before stop the timer, we compute the minutes spent with the timer
+
+            :return minutes_spent if the timer is started,
+                    otherwise return False
+        """
+        self.ensure_one()
```
Comment (tivisse) [inline addons/hr_timesheet/models/timer_mixin.py]:
``` python
def action_timer_stop(self):
    """ Stop the timer and return the spent minutes since it started
        :return minutes_spent if the timer is started,
                otherwise return False
    """
    self.ensure_one()
    if not self.timer_start:
        return False
    self.write({'timer_start': False, 'timer_pause': False})
    return self._compute_minutes_spent()
```

File: addons/sale_timesheet/models/account_move.py [hunk 18]
```diff
@@ -55,17 +81,17 @@ def create(self, vals_list):
...
-        for line in lines_to_process:
-            sale_line_delivery = line.sale_line_ids.filtered(lambda sol: sol.product_id.invoice_policy == 'delivery' and sol.product_id.service_type == 'timesheet')
-            if sale_line_delivery:
-                domain = self._timesheet_domain_get_invoiced_lines(sale_line_delivery)
-                timesheets = self.env['account.analytic.line'].search(domain).sudo()
-                timesheets.write({
-                    'timesheet_invoice_id': line.move_id.id,
-                })
-        return lines
+        # lines_to_process = lines.filtered(lambda line: line.move_id.type == 'out_invoice'
```
Comment (tivisse) [inline addons/sale_timesheet/models/account_move.py]: Commented code.

File: addons/sale_timesheet/models/product.py [hunk 19]
```diff
@@ -113,3 +113,7 @@ def _onchange_service_tracking(self):
             self.project_template_id = False
         elif self.service_tracking in ['task_in_project', 'project_only']:
             self.project_id = False
+
+    def is_delivered_timesheet(self):
+        """ Check if the product is a delivered timesheet """
+        return self.type == 'service' and self.service_policy == 'delivered_timesheet'
```
Comment (tivisse) [inline addons/sale_timesheet/models/product.py]: self.ensure_one() or even better, try to batch this ?

File: addons/hr_timesheet/views/hr_timesheet_views.xml (hunk 14 above)
Comment (xaviier12) [reply addons/hr_timesheet/views/hr_timesheet_views.xml]: In fact, the basic idea is, I must to only display this timer button in the menu ''My Timesheet", this condition was for this. But I have some problems with this condition in the kanban view, than I use a "magic context" as @LucasLefevre remarked it in a next commit. Perhaps, there is another solution, I will search this.

File: addons/sale_timesheet/models/sale_order.py [hunk 20]
```diff
@@ -146,6 +147,21 @@ def action_view_timesheet(self):
...
+
+            Add new parameter in this method, to invoice sale.order with a date. This date is used in sale_make_invoice_advance_inv into this module.
+
+            :param date
+
+            :return {account.move}: the invoices created
+        """
+        moves = super(SaleOrder, self)._create_invoices(grouped, final)
+
+        moves.search_timesheet_to_link(date)
```
Comment (tivisse) [inline addons/sale_timesheet/models/sale_order.py]: This method doesn't only search records.

It also write on them. So find a better method name ;)

File: addons/sale_timesheet/tests/test_sale_timesheet.py [hunk 21]
```diff
@@ -379,3 +384,138 @@ def test_timesheet_manual(self):
 
         self.assertFalse(timesheet1.timesheet_invoice_id, "The timesheet1 should not be linked to the invoice, even after invoice validation")
         self.assertFalse(timesheet2.timesheet_invoice_id, "The timesheet2 should not be linked to the invoice, even after invoice validation")
+
+    def test_timesheet_invoice(self):
```
Comment (tivisse) [inline addons/sale_timesheet/tests/test_sale_timesheet.py]: :boar: :boar: :boar: :boar: :boar: :boar: :boar: :boar: :boar: :boar: :boar: :boar:

File: addons/sale_timesheet/wizard/sale_make_invoice_advance.py [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+        elif self._context.get('active_ids', []):
+            order_lines = self.env['sale.order'].browse(self._context.get('active_ids')).mapped('order_line')
+            product_ids = order_lines.mapped('product_id')
+
+        if product_ids:
+            return any(product.is_delivered_timesheet for product in product_ids)
+        return False
+
+    date_invoice_timesheet = fields.Date(
+        string='Date invoice timesheet',
```
Comment (tivisse) [inline addons/sale_timesheet/wizard/sale_make_invoice_advance.py]: Date Invoice Timesheet

Comment (tivisse) [reply addons/sale_timesheet/wizard/sale_make_invoice_advance.py]: And find a more user friendly name ;)

File: addons/sale_timesheet/wizard/sale_make_invoice_advance.py [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+        default=fields.Date.today(),
+        required=False,
+        help="This field is used to invoice timesheets up to this date")
+    invoicing_timesheet_enabled = fields.Boolean(default=_default_invoicing_timesheet_enabled)
+
+    def create_invoices(self):
```
Comment (tivisse) [inline addons/sale_timesheet/wizard/sale_make_invoice_advance.py]: def action_generate_invoices ?

File: addons/sale_timesheet/wizard/sale_make_invoice_advance.py [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+        """ Override method from sale/wizard/sale_make_invoice_advance.py
+
+            When the user want to invoice the timesheets to the SO
+            up to a specific date then we need to recompute the
+            qty_to_invoice for each product_id in sale.order.line,
+            before creating the invoice.
+        """
+        sale_orders = self.env['sale.order'].browse(
```
Comment (tivisse) [inline addons/sale_timesheet/wizard/sale_make_invoice_advance.py]: You can store the ``order_ids`` on the wizard, and set it in the ``default_get``. That way, you don't have to make all those ``self._context.get('active_ids', [])``, and by the way, if you create this wizard in a python code, you just have to set 'order_id' in the ``vals`` dictionary instead of playing with ``with_context`` and hoping for the best ;)

File: addons/sale_timesheet/wizard/sale_make_invoice_advance.py [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+            self._context.get('active_ids', [])
+        )
+
+        if self.advance_payment_method == 'delivered':
```
Comment (tivisse) [inline addons/sale_timesheet/wizard/sale_make_invoice_advance.py]:
```python
if self.advance_payment_method == 'delivered' and self.invoicing_timesheet_enabled and self.date_invoice_timesheet:
    sale_orders.mapped('order_line').recompute_qty_to_invoice(self.date_invoice_timesheet)
    ....
```

File: addons/hr_timesheet/models/hr_timesheet.py (hunk 3 above)
Comment (tivisse) [reply addons/hr_timesheet/models/hr_timesheet.py]: Looks like a quick hack nope ?

File: addons/hr_timesheet/views/hr_timesheet_views.xml (hunk 12 above)
Comment (LucasLefevre) [reply addons/hr_timesheet/views/hr_timesheet_views.xml]: \* As truthy as `readonly="1"` ^^

File: addons/hr_timesheet/models/timer_mixin.py (hunk 17 above)
Comment (xaviier12) [reply addons/hr_timesheet/models/timer_mixin.py]: @tivisse Thanks, I rewrite this method, but I can't call the "_compute_minutes_spent" method after the "write" method because I need the value of 'timer_start' field before this "write". Therefore, I will let the local variable named 'minutes_spent'

File: addons/hr_timesheet/views/hr_timesheet_views.xml (hunk 15 above)
[the remark this answers is not in the document]
Comment (xaviier12) [reply addons/hr_timesheet/views/hr_timesheet_views.xml]: I use this to display the timer only in the 'My timesheet' page, because the 'My timesheet' and 'All timesheets' pages use the same views. 
I renamed this magic context to "my_timesheet_display_timer".

File: addons/hr_timesheet/models/hr_timesheet.py (hunk 3 above)
Comment (xaviier12) [reply addons/hr_timesheet/models/hr_timesheet.py]: I change this condition because it doesn't work the way I want it to. The idea is to check if the unit_amount is equal to 0 and minutes_spent < 1, then we have a timesheet with unit_amount field set on 0. And afterwards, we can invoice this timesheet, it's useless. To not have this, I initialize the minutes_spent to 1 (if the condition is True) and then we don't have the timesheet with unit_amount to 0 when we stop the timer.

Comment (LucasLefevre) [reply addons/hr_timesheet/models/hr_timesheet.py]: By looking at this a second time, the behavior is really strange.
For a brand new timesheet, 10s counts as a full minute.
But after, 10s counts as 0 minute.
:swimmer:
I would ask AJU again if she really wants this.

**Comment by xaviier12** [comment]
@tivisse Sorry, I just see, I forgot to squash the last commit (id: 884b415) with the commit with the same name (id: af100de).

**Comment by tivisse** [comment]
@robodoo r+ merge
